### ISSUE TITLE
java.util.function.Function with Message<> interface if return null not ignore message

### ISSUE THREAD

**Issue by sabbyanandan** [spring-cloud/spring-cloud-stream#2116]
@revkov commented on [Tue Feb 16 2021](https://github.com/spring-cloud/spring-cloud-dataflow/issues/4391)

I create java.util.function.Function processor:
```
@Bean 
public Function<Message<String>, Message<String>> ignoreHello() {
     return inputMessage -> {if (inputMessage.getPayload().equals("hello")) return null else return MessageBuilder.withPayload(inputMessage.getPayload).setHeader("status","OK");};
}
```

When i run it, on "hello" messages i get exeption:
```
Caused by: java.lang.NullPointerException: Cannot invoke "org.springframework.messaging.Message.getHeaders()" because "requestMessage" is null
	at org.springframework.messaging.core.GenericMessagingTemplate.sendTimeout(GenericMessagingTemplate.java:252) ~[spring-messaging-5.3.3.jar:5.3.3]
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:166) ~[spring-messaging-5.3.3.jar:5.3.3]
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47) ~[spring-messaging-5.3.3.jar:5.3.3]
	at org.springframework.messaging.core.AbstractDestinationResolvingMessagingTemplate.send(AbstractDestinationResolvingMessagingTemplate.java:72) ~[spring-messaging-5.3.3.jar:5.3.3]
	at org.springframework.cloud.stream.function.FunctionConfiguration$FunctionToDestinationBinder$1.doSendMessage(FunctionConfiguration.java:585) ~[spring-cloud-stream-3.1.0.jar:3.1.0]
	at org.springframework.cloud.stream.function.FunctionConfiguration$FunctionToDestinationBinder$1.handleMessageInternal(FunctionConfiguration.java:571) ~[spring-cloud-stream-3.1.0.jar:3.1.0]
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:56) ~[spring-integration-core-5.4.3.jar:5.4.3]
	... 32 common frames omitted
```

if i changed second argument from Message<String> to String it works fine. but i can not set custom headers:
```
@Bean
    public Function<Message<String>, String> ignoreHello() {
        return inputMessage -> {if (inputMessage.getPayload().equals("hello")) return null else return inputMessage.getPayload;};
}
```
I think this is bug. Please fix it, Maybe your need check returned result in both cases and no cast to Message<> interface, only full ignore.

**Release versions:**
<spring-cloud.version>2020.0.0</spring-cloud.version>
SCDF 2.7.2, stream 3.1.0

**Steps to reproduce:**
Create processor with functions on top and send message with body "hello"

**Comment by olegz** [comment]
This is actually fixed and released with 3.1.1

**Comment by Stexxen** [comment]
@olegz This [commit](https://github.com/spring-cloud/spring-cloud-stream/commit/31b851c17bface75c3cb228386de9df174ce80ed) to fix the NPE was created after 3.1.1 was released, so could we expect it in 3.1.2?
